### PR TITLE
Support scales as inputs in ResizeNearest

### PR DESCRIPTION
**Pull request by houseroad**
To address https://github.com/onnx/onnx/pull/1467

### CODE REVIEW

**Comment by houseroad** [comment]
There will be another PR to adjust the pytorch ONNX exporter to support new version Upsample

File: caffe2/operators/resize_op.cu [hunk 1]
```diff
@@ -79,6 +79,16 @@ bool ResizeNearestOp<float, CUDAContext>::RunOnDevice() {
   CAFFE_ENFORCE_EQ(4, inputDims.size());
   const int batch_size = X.dim32(0), num_channels = X.dim32(1),
             input_height = X.dim32(2), input_width = X.dim32(3);
+  if (InputSize() == 2) {
+    const auto& scales = Input(1);
+    CAFFE_ENFORCE_EQ(scales.ndim(), 1);
+    CAFFE_ENFORCE_EQ(scales.size(), 4);
+    float scales_data[4];
+    context_.CopyBytesToCPU(
```
Comment (BIT-silence) [inline caffe2/operators/resize_op.cu]:
```suggestion
    context_.CopyToCPU<float>(4, scales.data<float>(), scales_data);
```

File: caffe2/operators/resize_op.cu [hunk 2]
```diff
@@ -115,6 +125,16 @@ bool ResizeNearestGradientOp<float, CUDAContext>::RunOnDevice() {
             input_height = dY.dim32(2), input_width = dY.dim32(3);
   int output_height = X.dim32(2);
   int output_width = X.dim32(3);
+  if (InputSize() == 3) {
+    const auto& scales = Input(2);
+    CAFFE_ENFORCE_EQ(scales.ndim(), 1);
+    CAFFE_ENFORCE_EQ(scales.size(), 4);
+    float scales_data[4];
+    context_.CopyBytesToCPU(
```
Comment (BIT-silence) [inline caffe2/operators/resize_op.cu]: ditto

**Review by xiaomengy — APPROVED** [review]

File: caffe2/operators/resize_op.cu (hunk 1 above)
Comment (houseroad) [reply caffe2/operators/resize_op.cu]: Nice, thanks!

File: caffe2/operators/resize_op.cc [hunk 3]
```diff
@@ -151,19 +168,29 @@ output_width = floor(input_width * width_scale)
 output_height = floor(output_height * height_scale)
 )DOC")
     .Input(0, "X", "Input tensor")
+    .Input(1, "scales", // the hack to support onnx spec
+        "4D, Scales tensor, [1., 1., width_scale, height_scale]")
```
Comment (bddppq) [inline caffe2/operators/resize_op.cc]: why do we need to store the first two `1.`?

Comment (houseroad) [reply caffe2/operators/resize_op.cc]: I reserved it for the future... You are right, don't need that, let's make it aligned with argument version.

File: caffe2/python/operator_test/resize_op_test.py [hunk 4]
```diff
@@ -102,6 +102,98 @@ def ref(dY, X):
...
+                            output_height,
+                            output_width).astype(np.float32)
+        scales = np.array([height_scale, width_scale]).astype(np.float32)
+
+        op = core.CreateOperator(
+            "ResizeNearestGradient",
+            ["dY", "X", "scales"],
+            ["dX"],
+            width_scale=width_scale,
+            height_scale=height_scale,
```
Comment (d4l3k) [inline caffe2/python/operator_test/resize_op_test.py]: These scales should be omitted right?

Comment (houseroad) [reply caffe2/python/operator_test/resize_op_test.py]: Good catch, removed

**Pull request closed without merging**